### PR TITLE
Indirect - QENS Fitting Interfaces - Selection of discontinuous spectra ranges

### LINKED ISSUE #22064: Indirect - Fitting Interfaces - Unable to provide discontinuous list of spectra
At present it is not possible to specify a discontinuous list of spectra within the fitting interfaces; only a minimum and maximum spectrum can be selected.

### PR DESCRIPTION
**Pull request by brandonhewer**
It is currently not possible to select discontinuous spectra ranges in the QENS Fitting Interfaces (ConvFit, IqtFit, MSDFit, JumpFit).

This PR adds a spectrum selection widget, designed in the MVP pattern, and implements its use in the QENS fitting interfaces. This widget allows for the selection of both continuous and discontinuous selection (through spinners and a line edit widget respectively). It also provides the option to mask the bins of individual spectrum; note that currently the mask should always be the same across spectra, since the only available fitting mode is sequential.

**To test:**
1. Navigate to Interfaces > Indirect > Data Analysis.
2. For the MSDFit, IqtFit, ConvFit and JumpFit interfaces, provide the supplied test files at the the top of the interface (for ConvFit, also supply the resolution file).
3. For MSDFit select a Gaussian model in the 'Fit Type' drop-down menu. Add 1 Exponential and a Stretched Exponential in the IqtFit interface. Select One Lorentzian in the ConvFit drop-down menu. Select 'Use Delta Function' in the ConvFit function browser.
4. For IqtFit, select '0' as the StartX and '0.12' as the EndX.
4. For each interface, select a continuous fitting range by ensuring 'Range' is selected in the 'Fit Spectra' drop-down menu and then using the two spinners to select any continuous range.
5. Click the 'Run' button for each of the fitting interfaces and ensure a fit is performed, for each of the selected spectra, and the results displayed for each fit.
6. For each interface, select a discontinuous fitting range by ensuring 'String' is selected in the 'Fit Spectra' drop-down menu and then enter a valid range string (first try an invalid string, e.g. out of bounds spectrum, and ensure an error message is displayed to the user).
7. Click the 'Run' button for each of the fitting interfaces and ensure a fit is performed, for each of the selected spectra, and the results displayed for each fit.
8. Perform a fit with either a continuous or discontinuous fitting range in each of the QENS fitting interfaces, this time adding a mask string (even number of comma-separated real numbers). This region will be excluded when a fit is performed.

**Test Files:**
[QENSFitTestFiles.zip](https://github.com/mantidproject/mantid/files/2116400/QENSFitTestFiles.zip)

Fixes #22064. 

Does this update require release notes?
- [x] Yes
- [ ] No

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [x] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [x] Do changes function as described? Add comments below that describe the tests performed?
- [x] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by TWJubb** [comment]
Before looking into the code, I thought I'd first follow your tests. 

I really like the care taken to prevent bad ranges being entered, particularly the use of regular expressions. I also really like the fact that the range widget is blanked out during fitting. This prevents crashing which can happen with some of the other buttons.

The majority of the tests work fine, however there are one or two behaviours which could cause problems if somebody mistypes. I feel like a lot of these edge cases would be fixed by taking the user input range string and parsing it into comma separated ranges. For example "5,5,4,4,4,2,1,1,1" would be parsed to "1-2,4-5". I've separated the issues by tab below

# MSDFit

Except for the points below, all working fine. However I can't test spectrum ranges as the file only has a single spectrum.

- This ones a bit odd. If I load your file in, select Gaussian and run, then go back to the fit selection and put it back to None, then I can change the lower range spin box to "1" even though spectrum 1 doesn't exist.

- The "Plot Result" button produces two blank plots (after running the fit), is that supposed to be the case?

- This is probably outside the scope of this PR, but when I delete workspaces in the ADS with the GUI open, Mantid crashes.

# IQtFit

- When adding the stretched exponential, there is a checkbox with either "True" or "False" next to it. Is this a little bit confusing? A checkbox is pretty intuitive by itself and the text simply duplicates the information.

- The range spin boxes work perfectly.

- If I use a huge number (e.g. 100000000) in the upper end of a range string, mantid crashes.

- If I run a simple range "10-15" the fit is performed and the output appears in the ADS. If I then change the range string to "10-18" (i.e. out of range) the fit is still performed, however no output is produced. 

- If I use a reversed range "5-2" the error box says ""Error running algorithm". Compared to the warning from other bad strings such as spectra out of range, this is less informative.

- Stylistic point :  If I use comma separation I can make really long names in the ADS, as I can just write out a list of every spectra. Could it be possible to parse the input string into its least verbose form (i.e. collect any consecutive ranges using "-").

- If I use two ranges with a common spectra "1-5,5-10" the workspace for spectra 5 is created twice, is this intentional?

- I can repeat spectra through comma separation "1,1,1,1,1". Only the one workspace is created inside the group. However the name of the group includes the full range string with the duplicates. Again coming back to the stylistic point this would be corrected by parsing and simplifying the input string.

- I am able to add a trailing comma to strings "1,2,", this runs the fitting but does not produce the workspace (in other tabs this does not happen).

# ConvFit

- Again using a large range (1000000+) causes a freeze-up and crash.

- This time the trailing comma ("1,2,3,4,5,") produces a warning "Please correct the following: Spectra". I'm assuming it reads past the comma and thinks "" is a spectra.

- Minor stylistic point: Is the term "Mask Spectra" correctly understood by users? I would be less confused by "Mask spectra bins from fit" or something.

- Using repeated spectra "1,1,1,1,1" produces only a single workspace (as expected), but remove the red fitted curve from the embedded plot for all the repeats.

# JumpFit

- The range widget doesn't appear in this tab.

**Comment by brandonhewer** [comment]
Regarding some of the comments:

**MSDFit:**
This crash is outside of the scope of this PR (an ongoing issue), but identifying this has helped me write this down to ensure it is fixed in the near future.

The Plot Result option is likely not producing blank plots, but plots with only a single small point; only one spectrum is used in the fit.

**IqtFit:**
It is intentional that an output is produced for each repeated spectra.

**ConvFit:**
This is a great point about changing the name to make it more clear.

**JumpFit:**
The range selection shouldn't exist in this interface -- a spectrum is selected using the drop-down menu. The mask spectrum should still exist here.

Stylistic points are quite interesting here -- this is not something I'd have immediately considered, so thanks for the suggestions!

File: qt/scientific_interfaces/Indirect/IndirectFitData.h [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+  }
+  typename std::vector<T>::const_iterator end() const { return m_vec.end(); }
+  const T &operator[](std::size_t index) const { return m_vec[index]; }
+
+private:
+  std::string m_str;
+  std::vector<T> m_vec;
+};
+
+using Spectra = boost::variant<DiscontinuousSpectra<std::size_t>,
```
[the remark this answers is not in the document]
Comment (brandonhewer) [reply qt/scientific_interfaces/Indirect/IndirectFitData.h]: The variant here allows for two representations of spectra as the fastest operations performed on a discontinuous range (equivalent to a vector of spectrum) or a continuous range (equivalent to a pair defining the bounds) are not necessarily the same. There are circumstances in which it is more optimal to represent spectra as a pair (e.g. checking for out of bounds input, combining two spectra selections, etc.).

The discontinuous spectra is a template class in case of the fitting model being used in other contexts, where spectrum may be represented as a plain int or other type. It's not particularly necessary in this context.

File: Framework/CurveFitting/test/Algorithms/PlotPeakByLogValueTest.h [hunk 2]
```diff
@@ -615,6 +590,30 @@ class PlotPeakByLogValueTest : public CxxTest::TestSuite {
     AnalysisDataService::Instance().clear();
   }
 
+  void test_exclude_range() {
```
Comment (TWJubb) [inline Framework/CurveFitting/test/Algorithms/PlotPeakByLogValueTest.h]: A propositional test name i.e. `test_excluding_range_does_some_thing` makes it clearer what the test is testing. It appears this one just simply checks that you can enter an excluded range and the algorithm still runs. Could there be more tests on the actual results of the fit? Also, are there any fail cases that need to be checked like giving an invalid range?

File: Framework/CurveFitting/test/Algorithms/PlotPeakByLogValueTest.h [hunk 3]
```diff
@@ -615,6 +590,30 @@ class PlotPeakByLogValueTest : public CxxTest::TestSuite {
...
+    alg.setPropertyValue("Input", "InputWS,i0");
+    alg.setPropertyValue("Exclude", "-0.5, 0.5");
+    alg.setPropertyValue("OutputWorkspace", "PlotPeakResult");
+    alg.setProperty("CreateOutput", true);
+    alg.setPropertyValue("Function", "name=FlatBackground,A0=2");
+    alg.setPropertyValue("MaxIterations", "50");
+    alg.execute();
+
+    TS_ASSERT(alg.isExecuted());
+    AnalysisDataService::Instance().remove("InputWS");
```
Comment (TWJubb) [inline Framework/CurveFitting/test/Algorithms/PlotPeakByLogValueTest.h]: Would it be safer to use `.clear()` here? Just in case there are some workspaces created in the algorithm or some left hanging around from previous tests?

File: qt/scientific_interfaces/Indirect/IndirectFitOutput.cpp [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+    auto &values = extractOrAddDefault(parameters, inputData);
+    return [&](std::size_t index, std::size_t spectrum) {
+      values[spectrum] = extractRowFromTable(index);
+    };
+  };
+  applyEnumeratedData(extract, fitDataBegin, fitDataEnd);
+}
+
+template <typename Map2D, typename T, typename U, typename Value>
+Value findOr(const Map2D &map, const T &firstKey, const U &secondKey,
```
Comment (TWJubb) [inline qt/scientific_interfaces/Indirect/IndirectFitOutput.cpp]: I struggle to understand what this function does from the name (you have another function called `extractOrAddDefault` so maybe this one could be `findOrAddDefault`)...

Just out of interest, is there a way to make this recursive so it would apply to arbitrarily nested maps?

File: qt/scientific_interfaces/Indirect/IndirectFitOutput.cpp [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+
+WorkspaceGroup_sptr IndirectFitOutput::getLastResultGroup() const {
+  return m_resultGroup.lock();
+}
+
+void IndirectFitOutput::mapParameterNames(
+    const std::unordered_map<std::string, std::string> &parameterNameChanges,
+    const FitDataIterator &fitDataBegin, const FitDataIterator &fitDataEnd) {
+  for (auto it = fitDataBegin; it < fitDataEnd; ++it) {
+    auto &parameterValues = m_parameters[it->get()];
```
Comment (TWJubb) [inline qt/scientific_interfaces/Indirect/IndirectFitOutput.cpp]: Is there any danger here that if the `IndirectFitData` does not already exists in `m_parameters` that a key is added for it, and then for example the `isSpectrumFit` method gives the wrong result..

**Review by TWJubb — CHANGES_REQUESTED** [review]

File: qt/scientific_interfaces/Indirect/IndirectFitData.cpp [hunk 6]
```diff
@@ -0,0 +1,299 @@
...
+  }
+
+  Spectra operator()(const Spectra &spectra1, const Spectra &spectra2) const {
+    return DiscontinuousSpectra<std::size_t>(
+        boost::apply_visitor(SpectraToString(), spectra1) + "," +
+        boost::apply_visitor(SpectraToString(), spectra2));
+  }
+};
+
+struct GetSpectrum : boost::static_visitor<std::size_t> {
```
[the remark this answers is not in the document]
Comment (brandonhewer) [reply qt/scientific_interfaces/Indirect/IndirectFitData.cpp]: To use the apply_visitor function from boost, you must inherit from boost::static_visitor. This makes sense to represent as a struct as it is functor-like, meaning it is sensible for public to be the default protection-level (and remains consistent with stl and boost).

File: qt/scientific_interfaces/Indirect/IndirectFitData.cpp [hunk 7]
```diff
@@ -0,0 +1,299 @@
...
+  }
+}
+
+void IndirectFitData::setStartX(double startX, std::size_t spectrum) {
+  const auto range = m_ranges.find(spectrum);
+  if (range != m_ranges.end())
+    range->second.first = startX;
+  else
+    m_ranges[spectrum] =
+        std::make_pair(startX, m_workspace.lock()->x(0).back());
```
[the remark this answers is not in the document]
Comment (brandonhewer) [reply qt/scientific_interfaces/Indirect/IndirectFitData.cpp]: This is absolutely true! Thanks for spotting this. It is possible for lock here to return a nullptr.

**Review by TWJubb — CHANGES_REQUESTED** [review]
Your code is very self documenting; but it does have very few comments anywhere. I do feel it would be a good idea to have comments in the header files at the top just to give a brief account of the responsibilities of the class. Particularly for `IndirectFitData`, `IndirectFitOutput` and `IndirectFittingModel`. 

You have used a lot of non-trivial language features and I think briefly explaining your intent with some of these would go a long way. E.g. a comment on the use of `Spectra` : "Use variant to allow discontinuous ranges but maintain speed for continuous ranges" helps me understand the `IndirectFitData` much quicker.

File: qt/scientific_interfaces/Indirect/IndirectFitOutput.cpp (hunk 4 above)
Comment (brandonhewer) [reply qt/scientific_interfaces/Indirect/IndirectFitOutput.cpp]: This method doesn't add to the map, unlike extractOrAddDefault, so probably shouldn't be called findOrAddDefault. It was modelled after boost::optional's get_value_or -- it may however still be best to rename.

Comment (TWJubb) [reply qt/scientific_interfaces/Indirect/IndirectFitOutput.cpp]: Ok fair enough. If it makes sense to people familiar with boost then this isn't an issue

File: qt/scientific_interfaces/Indirect/IndirectSpectrumSelectionPresenter.h [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+
+private:
+  void setSpectraRange(std::size_t minimum, std::size_t maximum);
+  void setModelSpectra(const Spectra &spectra);
+
+  UserInputValidator validateSpectraString();
+  UserInputValidator &validateSpectraString(UserInputValidator &validator);
+  UserInputValidator validateMaskBinsString();
+
+  IndirectFittingModel *m_model;
```
Comment (TWJubb) [inline qt/scientific_interfaces/Indirect/IndirectSpectrumSelectionPresenter.h]: Is there a reason that `m_model` is not a `unique_ptr` or some other kind of smart pointer?

File: qt/scientific_interfaces/Indirect/IndirectFitAnalysisTab.cpp [hunk 9]
```diff
@@ -20,139 +20,20 @@
...
-    std::sort(functions1.begin(), functions1.end(), functionNameComparator);
-    std::sort(functions2.begin(), functions2.end(), functionNameComparator);
-
-    for (auto i = 0u; i < functions1.size(); ++i) {
-      if (!equivalentFunctions(functions1[i], functions2[i]))
-        return false;
+void updateParameters(
+    IFunction_sptr function,
+    const std::unordered_map<std::string, ParameterValue> &parameters) {
+  for (auto i = 0u; i < function->nParams(); ++i) {
```
Comment (TWJubb) [inline qt/scientific_interfaces/Indirect/IndirectFitAnalysisTab.cpp]: Is it possible here to use `for(auto && name : function->getParameterNames())`, and use the overloaded `setParameter` which takes the name?

Actually, not sure if there is an appropriate `setError` function for this case.

File: qt/scientific_interfaces/Indirect/IndirectSpectrumSelectionPresenter.h (hunk 8 above)
Comment (brandonhewer) [reply qt/scientific_interfaces/Indirect/IndirectSpectrumSelectionPresenter.h]: This is a non-owning pointer -- this is important as the memory of the model should only be owned from a single place. This presenter does not need to worry about cleaning up the memory associated to this pointer.

File: qt/scientific_interfaces/Indirect/IndirectFitAnalysisTab.cpp (hunk 9 above)
Comment (brandonhewer) [reply qt/scientific_interfaces/Indirect/IndirectFitAnalysisTab.cpp]: This is a possibility -- the problem is that the setParameter method which takes a name rather than an index, goes onto find the index itself -- this would be slower since I would need to find the index of each parameter on every iteration.

File: sh.exe.stackdump [hunk 10]
```diff
@@ -0,0 +1,28 @@
+Exception: STATUS_STACK_OVERFLOW at rip=7FFF8F7CAA67
```
Comment (TWJubb) [inline sh.exe.stackdump]: Did you intend to include this file?

**Review by TWJubb — APPROVED** [review]
